Add ILogger bytecode instrumentation (#1262)
open-telemetry/opentelemetry-dotnet-instrumentation: docs/config.md

| `HttpClient` | [System.Net.Http.HttpClient](https://docs.microsoft.com/dotnet/api/system.net.http.httpclient) and [System.Net.HttpWebRequest](https://docs.microsoft.com/dotnet/api/system.net.httpwebrequest) | * | source |
| `NetRuntime` | [OpenTelemetry.Instrumentation.Runtime](https://www.nuget.org/packages/OpenTelemetry.Instrumentation.Runtime) | * | source |

## Logs

Logs from ASP.NET Core apps could be captured using two options. First option is
to use HostingStartup and the second option is to rely on CLR Profiler
Instrumentation. Recommended option for ASP.NET Core is to use HostingStartup.
For other .NET Core apps logs could be collected only through .NET CLR Profiler

> Instrumentation.

Reviewer: This could be a note in the "Instrumented logging libraries and frameworks" section.

> Logs from ASP.NET Core apps could be captured using two options. First option is
> to use HostingStartup and the second option is to rely on CLR Profiler
> Instrumentation. Recommended option for ASP.NET Core is to use HostingStartup.
> For other .NET Core apps logs could be collected only through .NET CLR Profiler
> Instrumentation.

Reviewer 2: ```suggestion
Logs from ASP.NET Core apps can be captured in two ways: The first option is
to use `HostingStartup`, and the second is to rely on CLR Profiler Instrumentation. 
For ASP.NET Core, use `HostingStartup`. For other .NET Core apps, you can only
collect logs through .NET CLR Profiler Instrumentation.
```

Author: If both are present `HostingStartup` adds the logger integration. There is no harm in having both of them.

### Enable Logs using HostingStartup

| Environment variable | Required value |
|-|-|

> | `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES` | `OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper` |

Reviewer: I see this as  "source instrumentation" for `ILogger` that requires additionally setting `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES`.

Maybe we could try to describe it as such in a `Instrumented logging libraries and frameworks` section?

### Enable Logs using .NET CLR Profiler

Please follow [instructions](config.md#net-clr-profiler) here to collect logs

> using .NET CLR Profiler.

Reviewer: I see this PR as adding "bytecode instrumentation" for `ILogger` 

Maybe we could try to describe it as such in a `Instrumented logging libraries and frameworks` section?

Author: There are no additional instrumentation options like traces or metrics, hence placed in a different section. Even if we add `Instrumented logging libraries and frameworks`, there will be only `AspNet` option. 

If I follow you correct, `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES` should get to source instrumentation and  .NET CLR Profiler part will move to `Instrumented logging frameworks` section. Is that correct?

Reviewer: I think it may be easier if we merge this and I will try to create a PR 😋 At least I think it will be easier and faster for me 😉

> ### Logs Parse and Format Options

Reviewer 3: This heading reads a little awkwardly for me. Maybe it should be something like the following.

```suggestion
### Parsing and Format Options for Logs
```

These options could be used with both HostingStartup and .NET CLR Profiler logs
enablement.

| Environment variable | Description | Default value |
|-|-|-|
| `OTEL_DOTNET_AUTO_LOGS_PARSE_STATE_VALUES` | Whether the log state should be parsed. | `false` |

> | `OTEL_DOTNET_AUTO_LOGS_INCLUDE_FORMATTED_MESSAGE` | Whether the log state should be formatted. | `false` |

Reviewer: How about moving these to the `Exporters` section as this looks like a common config for the log exporters?
